File: go/pkg/hey/extenzions.go
package hey

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"regexp"
	"strconv"
	"time"

	"github.com/basecamp/hey-sdk/go/pkg/generated"
)

// ExtenzionsService handles email extenzion operations.
//
// Extenzions allow custom email addresses on custom-domain HEY accounts
// (e.g., sales@yourdomain.com). Create and Update take form-encoded requests but post to the
// .json path, so a current server answers the written extenzion; a server without the JSON
// branch redirects instead and hands nothing back. Delete is a plain JSON operation.
type ExtenzionsService struct {
	client *Client
}

// NewExtenzionsService creates a new ExtenzionsService.
func NewExtenzionsService(client *Client) *ExtenzionsService {
	return &ExtenzionsService{client: client}
}

// Extenzion represents an email extenzion.
//
// The id is the extenzion's contact id — the same id the write endpoints take.
type Extenzion struct {
	ID     int64
	Name   string
	AppURL string
}

// CreateExtenzionParams contains the parameters for creating an extenzion.
type CreateExtenzionParams struct {
	// Name is the extenzion name (e.g., "sales" becomes sales@yourdomain.com).
	Name string
	// Members is a list of member email addresses.
	Members []string
}

// UpdateExtenzionParams contains the parameters for updating an extenzion.
type UpdateExtenzionParams struct {
	// Name is the new extenzion name. Empty string means no change.
	Name string
	// Members is the new list of member email addresses. Replaces all existing members.
	// nil means no change.
	Members []string
}

// List returns the extenzions on the account.
//
// This reads the navigation payload rather than scraping the extenzions page, so it carries
// only what navigation carries: each extenzion's name and its contact URL.
func (s *ExtenzionsService) List(ctx context.Context) (result []Extenzion, err error) {
	op := OperationInfo{
		Service: "Extenzions", Operation: "ListExtenzions",
		ResourceType: "extenzion", IsMutation: false,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		navigation, nerr := s.client.Identity().GetNavigation(ctx)
		if nerr != nil {
			return nerr
		}
		result = extenzionsFromNavigation(navigation)
		return nil
	})
	return result, err
}

// Create creates a new extenzion and returns it. A server without the JSON create branch
// hands nothing back, and the result is then nil.
func (s *ExtenzionsService) Create(ctx context.Context, accountID int64, params CreateExtenzionParams) (extenzion *Extenzion, err error) {
	op := OperationInfo{
		Service: "Extenzions", Operation: "CreateExtenzion",
		ResourceType: "extenzion", IsMutation: true,
	}
	if gater, ok := s.client.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	values := url.Values{}
	values.Set("extenzion[name]", params.Name)
	for _, m := range params.Members {
		values.Add("extenzion[members][]", m)
	}

	resp, err := s.client.PostForm(ctx, fmt.Sprintf("/accounts/%d/domains/extenzions.json", accountID), values)
	if err != nil {
		return nil, err
	}
	return extenzionFromFormResponse(resp)
}

// Update updates an existing extenzion and returns it. A server without the JSON update
// branch hands nothing back, and the result is then nil.
func (s *ExtenzionsService) Update(ctx context.Context, accountID int64, extID int64, params UpdateExtenzionParams) (extenzion *Extenzion, err error) {
	op := OperationInfo{
		Service: "Extenzions", Operation: "UpdateExtenzion",
		ResourceType: "extenzion", IsMutation: true, ResourceID: extID,
	}
	if gater, ok := s.client.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	values := url.Values{}
	if params.Name != "" {
		values.Set("extenzion[name]", params.Name)
	}
	if params.Members != nil {
		for _, m := range params.Members {
			values.Add("extenzion[members][]", m)
		}
	}

	resp, err := s.client.PatchForm(ctx, fmt.Sprintf("/accounts/%d/domains/extenzions/%d.json", accountID, extID), values)
	if err != nil {
		return nil, err
	}
	return extenzionFromFormResponse(resp)
}

// Delete deletes an extenzion.
func (s *ExtenzionsService) Delete(ctx context.Context, accountID int64, extID int64) (err error) {
	op := OperationInfo{
		Service: "Extenzions", Operation: "DeleteExtenzion",
		ResourceType: "extenzion", IsMutation: true, ResourceID: extID,
	}
	if gater, ok := s.client.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	resp, err := s.client.genClient().DeleteExtenzionWithResponse(ctx, accountID, extID)
	if err != nil {
		return err
	}
	return CheckResponse(resp.HTTPResponse)
}

// extenzionFromFormResponse reads the extenzion a JSON write answers with. A server without
// the JSON branch redirects to the extenzions page instead, which leaves nothing to read.
//
// The payload's own id is the Extenzion record's, while every write endpoint takes the
// contact's — so the id comes out of app_url, the way List reads it out of navigation.
func extenzionFromFormResponse(resp *FormResponse) (*Extenzion, error) {
	if len(resp.Body) == 0 {
		return nil, nil
	}

	var payload struct {
		Name   string `json:"name"`
		AppURL string `json:"app_url"`
	}
	if err := json.Unmarshal([]byte(resp.Body), &payload); err != nil {
		return nil, fmt.Errorf("failed to decode the extension: %w", err)
	}

	id, err := contactIDFromURL(payload.AppURL)
	if err != nil {
		return nil, err
	}
	return &Extenzion{ID: id, Name: payload.Name, AppURL: payload.AppURL}, nil
}

// contactIDFromURL pulls the contact id out of a contact URL, e.g. /contacts/4821.
func contactIDFromURL(contactURL string) (int64, error) {
	match := contactPathRe.FindStringSubmatch(contactURL)
	if match == nil {
		return 0, fmt.Errorf("no contact id in %q", contactURL)
	}
	return strconv.ParseInt(match[1], 10, 64)
}

// contactPathRe matches the contact URLs navigation gives each extenzion, e.g. /contacts/4821
var contactPathRe = regexp.MustCompile(`/contacts/(\d+)`)

// extenzionsFromNavigation pulls the extenzions out of the "Extensions" navigation group.
// The group's first entry is the "All Extensions"/"Manage Extensions" link, which has no
// contact URL and so falls out on its own.
func extenzionsFromNavigation(navigation *generated.NavigationResponse) []Extenzion {
	if navigation == nil {
		return nil
	}

	var extenzions []Extenzion
	for _, item := range navigation.Items {
		if item.Title != navigationExtenzionsTitle {
			continue
		}
		for _, entry := range item.MenuItems {
			match := contactPathRe.FindStringSubmatch(entry.AppUrl)
			if match == nil {
				continue
			}
			id, err := strconv.ParseInt(match[1], 10, 64)
			if err != nil {
				continue
			}
			extenzions = append(extenzions, Extenzion{ID: id, Name: entry.Title, AppURL: entry.AppUrl})
		}
	}
	return extenzions
}

// navigationExtenzionsTitle is the title HEY gives the extensions group in the navigation payload.
const navigationExtenzionsTitle = "Extensions"

File: rust/hey-sdk/src/services/extenzions.rs
//! Email extenzions — the custom addresses a custom-domain account carries, such as
//! `sales@yourdomain.com`.
//!
//! [`Extenzions::create`] and [`Extenzions::update`] post a form to the `.json` path, so a
//! current server answers the written extenzion while one without the JSON branch redirects
//! and hands nothing back. The generated [`Extenzions::delete`] is a plain JSON operation.

use std::borrow::Cow;

use reqwest::Method;

use crate::client::Response;
use crate::error::Error;
use crate::form::FormResponse;
use crate::generated::routes;
use crate::generated::types::{Extenzion as ExtenzionPayload, NavigationItem, NavigationResponse};
use crate::observability::OperationInfo;
use crate::services::write_info;

pub use crate::generated::services::extenzions::*;

/// The title HEY gives the extensions group in the navigation payload.
const NAVIGATION_GROUP: &str = "Extensions";

/// What a contact URL puts the contact's id after, as in `/contacts/4821`.
const CONTACT_PATH: &str = "/contacts/";

/// One email extenzion.
///
/// The id is the extenzion's *contact* id — the one every write endpoint takes, and the one
/// its `app_url` carries. The id a JSON write answers with belongs to the Extenzion record
/// instead, which no endpoint takes.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Extenzion {
    pub id: i64,
    pub name: String,
    pub app_url: String,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct CreateExtenzionParams {
    /// The extenzion name: "sales" becomes `sales@yourdomain.com`.
    pub name: String,
    /// The member email addresses.
    pub members: Vec<String>,
}

/// A partial revision: `None` leaves a field as it is.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct UpdateExtenzionParams {
    /// An empty name is no name, and is left off the wire like a `None` one.
    pub name: Option<String>,
    /// The whole membership, which replaces what the extenzion had rather than adding to
    /// it.
    pub members: Option<Vec<String>>,
}

impl<'a> Extenzions<'a> {
    /// The extenzions on the account.
    ///
    /// This reads the navigation payload rather than scraping the extenzions page, so it
    /// carries only what navigation carries: each extenzion's name and its contact URL.
    /// It is one operation, `Extenzions.ListExtenzions`, rather than the identity read it
    /// happens to be built on — which is why it sends the navigation route itself instead
    /// of going through [`Identity::get_navigation`](crate::services::identity::Identity::get_navigation).
    pub async fn list(&self) -> Result<Vec<Extenzion>, Error> {
        let mut operation = self.client().operation(&routes::GET_NAVIGATION, &[]);
        operation.info(OperationInfo {
            service: Cow::Borrowed("Extenzions"),
            operation: Cow::Borrowed("ListExtenzions"),
            resource_type: Cow::Borrowed("extenzion"),
            is_mutation: false,
            resource_id: None,
        });
        let navigation: NavigationResponse = self.client().send(operation).await?;
        extenzions_from_navigation(&navigation)
    }

    /// Creates an extenzion and answers it. A server without the JSON create branch hands
    /// nothing back, and the answer is then `None`.
    pub async fn create(
        &self,
        account_id: i64,
        params: &CreateExtenzionParams,
    ) -> Result<Option<Extenzion>, Error> {
        let mut fields = vec![("extenzion[name]", params.name.as_str())];
        for member in &params.members {
            fields.push(("extenzion[members][]", member));
        }

        let mut operation = self.client().form(
            Method::POST,
            &format!("/accounts/{account_id}/domains/extenzions.json"),
        )?;
        operation.info(write_info(
            "Extenzions",
            "CreateExtenzion",
            "extenzion",
            None,
        ));
        operation.form(&fields);
        extenzion_from_form_response(&self.client().execute(operation).await?)
    }

    /// Revises an extenzion and answers it. The id is the extenzion's contact id. A server
    /// without the JSON update branch hands nothing back, and the answer is then `None`.
    pub async fn update(
        &self,
        account_id: i64,
        extenzion_id: i64,
        params: &UpdateExtenzionParams,
    ) -> Result<Option<Extenzion>, Error> {
        let mut fields = Vec::new();
        if let Some(name) = &params.name
            && !name.is_empty()
        {
            fields.push(("extenzion[name]", name.as_str()));
        }
        if let Some(members) = &params.members {
            for member in members {
                fields.push(("extenzion[members][]", member));
            }
        }

        let mut operation = self.client().form(
            Method::PATCH,
            &format!("/accounts/{account_id}/domains/extenzions/{extenzion_id}.json"),
        )?;
        operation.info(write_info(
            "Extenzions",
            "UpdateExtenzion",
            "extenzion",
            Some(extenzion_id),
        ));
        operation.form(&fields);
        extenzion_from_form_response(&self.client().execute(operation).await?)
    }
}

/// The extenzions in navigation's "Extensions" group. The group leads with the "All
/// Extensions" link, which names no contact and so falls out on its own.
fn extenzions_from_navigation(navigation: &NavigationResponse) -> Result<Vec<Extenzion>, Error> {
    navigation
        .items
        .iter()
        .flatten()
        .filter(|item| item.title.as_deref() == Some(NAVIGATION_GROUP))
        .flat_map(|group| group.menu_items.iter().flatten())
        .filter_map(|entry| listed_extenzion(entry).transpose())
        .collect()
}

fn listed_extenzion(entry: &NavigationItem) -> Result<Option<Extenzion>, Error> {
    let app_url = entry.app_url.clone().unwrap_or_default();
    Ok(contact_id_from_url(&app_url)?.map(|id| Extenzion {
        id,
        name: entry.title.clone().unwrap_or_default(),
        app_url,
    }))
}

/// The extenzion a JSON write answered with. A server without the JSON branch redirects to
/// the extenzions page instead, which leaves nothing to read.
fn extenzion_from_form_response(answered: &Response) -> Result<Option<Extenzion>, Error> {
    let written = FormResponse::new(answered);
    if written.body.is_empty() {
        Ok(None)
    } else {
        let payload: ExtenzionPayload = serde_json::from_str(&written.body)?;
        let app_url = payload.app_url.unwrap_or_default();
        let id = contact_id_from_url(&app_url)?
            .ok_or_else(|| Error::api(0, format!("no contact id in {app_url:?}")))?;
        Ok(Some(Extenzion {
            id,
            name: payload.name.unwrap_or_default(),
            app_url,
        }))
    }
}

/// The contact id a contact URL carries, as Go's `/contacts/(\d+)` reads it.
///
/// A URL naming no contact at all answers `None` — navigation's "All Extensions" link is
/// one, and it is meant to fall out of the list. A URL that names one the SDK cannot read
/// is a failure instead of another such link: dropping it would hide an extenzion the
/// caller would then never hear about.
fn contact_id_from_url(contact_url: &str) -> Result<Option<i64>, Error> {
    for (at, _) in contact_url.match_indices(CONTACT_PATH) {
        let digits = digit_run(&contact_url[at + CONTACT_PATH.len()..]);
        if !digits.is_empty() {
            return digits.parse().map(Some).map_err(|error| {
                Error::api(
                    0,
                    format!("contact id in {contact_url:?} is not a number: {error}"),
                )
            });
        }
    }
    Ok(None)
}

fn digit_run(text: &str) -> String {
    text.chars().take_while(char::is_ascii_digit).collect()
}
